File: src/priv/bajdcc/LALR1/interpret/module/ModuleNet.java
package priv.bajdcc.LALR1.interpret.module;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import org.apache.log4j.Logger;
import priv.bajdcc.LALR1.grammar.Grammar;
import priv.bajdcc.LALR1.grammar.runtime.*;
import priv.bajdcc.LALR1.grammar.runtime.data.RuntimeArray;
import priv.bajdcc.LALR1.grammar.runtime.data.RuntimeMap;
import priv.bajdcc.LALR1.interpret.module.net.ModuleNetClient;
import priv.bajdcc.LALR1.interpret.module.net.ModuleNetServer;
import priv.bajdcc.util.ResourceLoader;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.net.URL;
import java.net.URLConnection;
import java.util.List;

/**
 * 【模块】网络
 *
 * @author bajdcc
 */
public class ModuleNet implements IInterpreterModule {

	private static Logger logger = Logger.getLogger("net");
	private static ModuleNet instance = new ModuleNet();

	private static final int MSG_QUERY_TIME = 15;
	private static final int MSG_SEND_TIME = 5;
	private RuntimeCodePage runtimeCodePage;
	private ModuleNetServer server;
	private ModuleNetClient client;
	private String lastError = "";

	public static ModuleNet getInstance() {
		return instance;
	}

	@Override
	public String getModuleName() {
		return "sys.net";
	}

	@Override
	public RuntimeCodePage getCodePage() throws Exception {
		if (runtimeCodePage != null)
			return runtimeCodePage;

		String base = ResourceLoader.load(getClass());

		Grammar grammar = new Grammar(base);
		RuntimeCodePage page = grammar.getCodePage();
		IRuntimeDebugInfo info = page.getInfo();
		buildRemoteMethods(info);

		return runtimeCodePage = page;
	}

	private static RuntimeObject parseJsonSafe(String text) {
		try {
			Object o = JSON.parseObject(text);
			return parseInternal(o);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return null;
	}

	private static RuntimeObject parseJson(String text) {
		try {
			Object o = JSON.parseObject(text);
			return parseInternal(o);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return new RuntimeObject("json - parse error");
	}

	private void buildRemoteMethods(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_net_get", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "HTTP GET";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				String text = "";
				String txt = String.valueOf(args.get(0).getObj());
				logger.debug("Request url: " + txt);
				try {
					URL url = new URL(txt);
					URLConnection urlConnection = url.openConnection(); // 打开连接
					BufferedReader br = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "utf-8")); // 获取输入流
					String line = null;
					StringBuilder sb = new StringBuilder();
					while ((line = br.readLine()) != null) {
						sb.append(line).append("\n");
					}
					text = sb.toString();
					br.close();
				} catch (Exception e) {
					e.printStackTrace();
				}
				return new RuntimeObject(text);
			}
		});
		info.addExternalFunc("g_net_get_json", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "HTTP GET - json";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				String text = "";
				String txt = String.valueOf(args.get(0).getObj());
				logger.debug("Request url(json): " + txt);
				try {
					URL url = new URL(txt);
					URLConnection urlConnection = url.openConnection(); // 打开连接
					BufferedReader br = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "utf-8")); // 获取输入流
					String line = null;
					StringBuilder sb = new StringBuilder();
					while ((line = br.readLine()) != null) {
						sb.append(line).append("\n");
					}
					text = sb.toString();
					br.close();
				} catch (Exception e) {
					e.printStackTrace();
				}
				return parseJson(text);
			}
		});
		// -----------------------------------
		// server
		info.addExternalFunc("g_net_msg_create_server_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG SERVER";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kInt};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getServer() != null || getClient() != null)
					return new RuntimeObject(false);
				int port = ((BigInteger) args.get(0).getObj()).intValue();
				setServer(new ModuleNetServer(port));
				getServer().start();
				return new RuntimeObject(true);
			}
		});
		info.addExternalFunc("g_net_msg_shutdown_server", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG SERVER SHUTDOWN";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getServer() == null)
					return new RuntimeObject(false);
				getServer().exit();
				return new RuntimeObject(true);
			}
		});
		info.addExternalFunc("g_net_msg_get_server_status", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG SERVER GET STATUS";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getServer() != null) {
					ModuleNetServer.Status s = getServer().getStatus();
					if (s == ModuleNetServer.Status.ERROR) {
						lastError = getServer().getError();
						setServer(null);
					}
					status.getService().getProcessService().sleep(status.getPid(), MSG_QUERY_TIME);
					return new RuntimeObject(BigInteger.valueOf(s.ordinal()));
				}
				return new RuntimeObject(BigInteger.valueOf(ModuleNetServer.Status.NULL.ordinal()));
			}
		});
		// server
		// -----------------------------------
		// client
		info.addExternalFunc("g_net_msg_create_client_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG CLIENT";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getServer() != null || getClient() != null)
					return new RuntimeObject(false);
				String addr = String.valueOf(args.get(0).getObj());
				setClient(new ModuleNetClient(addr));
				getClient().start();
				return new RuntimeObject(true);
			}
		});
		info.addExternalFunc("g_net_msg_shutdown_client", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG CLIENT SHUTDOWN";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getClient() == null)
					return new RuntimeObject(false);
				getClient().exit();
				return new RuntimeObject(true);
			}
		});
		info.addExternalFunc("g_net_msg_get_client_status", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG CLIENT GET STATUS";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getClient() != null) {
					ModuleNetClient.Status s = getClient().getStatus();
					if (s == ModuleNetClient.Status.ERROR) {
						lastError = getClient().getError();
						setClient(null);
					}
					status.getService().getProcessService().sleep(status.getPid(), MSG_QUERY_TIME);
					return new RuntimeObject(BigInteger.valueOf(s.ordinal()));
				}
				return new RuntimeObject(BigInteger.valueOf(ModuleNetClient.Status.NULL.ordinal()));
			}
		});
		info.addExternalFunc("g_net_msg_client_send", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG CLIENT SEND";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				if (getClient() != null) {
					ModuleNetClient.Status s = getClient().getStatus();
					if (s == ModuleNetClient.Status.RUNNING) {
						status.getService().getProcessService().sleep(status.getPid(), MSG_SEND_TIME);
						getClient().send(String.valueOf(args.get(0).getObj()));
						return new RuntimeObject(true);
					}
				}
				return new RuntimeObject(false);
			}
		});
		// client
		// -----------------------------------
		info.addExternalFunc("g_net_msg_get_error", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG SERVER GET ERROR";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				return new RuntimeObject(lastError);
			}
		});
		info.addExternalFunc("g_net_msg_get_server_msg", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG SERVER GET MESSAGE";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				return new RuntimeObject(getServer().getMessage());
			}
		});
		info.addExternalFunc("g_net_msg_get_client_msg", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "MSG CLIENT GET MESSAGE";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				return new RuntimeObject(getClient().getMessage());
			}
		});
		info.addExternalFunc("g_net_parse_json", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "Parse json";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) {
				return parseJsonSafe(String.valueOf(args.get(0).getObj()));
			}
		});
	}

	private static RuntimeObject parseInternal(Object o) {
		if (o instanceof JSONObject) {
			JSONObject obj = (JSONObject) o;
			RuntimeMap map = new RuntimeMap();
			obj.forEach((key, value) -> {
				map.put(key, parseInternal(value));
			});
			return new RuntimeObject(map);
		} else if (o instanceof JSONArray) {
			JSONArray obj = (JSONArray) o;
			RuntimeArray arr = new RuntimeArray();
			obj.forEach((key) -> {
				arr.add(parseInternal(key));
			});
			return new RuntimeObject(arr);
		} else {
			return new RuntimeObject(o);
		}
	}

	public ModuleNetServer getServer() {
		return server;
	}

	public void setServer(ModuleNetServer server) {
		this.server = server;
	}

	public ModuleNetClient getClient() {
		return client;
	}

	public void setClient(ModuleNetClient client) {
		this.client = client;
	}
}

File: src/main/kotlin/com/bajdcc/LALR1/interpret/module/ModuleNet.kt
package com.bajdcc.LALR1.interpret.module

import com.alibaba.fastjson.JSON
import com.alibaba.fastjson.JSONArray
import com.alibaba.fastjson.JSONObject
import com.bajdcc.LALR1.grammar.Grammar
import com.bajdcc.LALR1.grammar.runtime.*
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeArray
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeMap
import com.bajdcc.LALR1.interpret.module.api.ModuleNetWebApi
import com.bajdcc.LALR1.interpret.module.net.ModuleNetClient
import com.bajdcc.LALR1.interpret.module.net.ModuleNetServer
import com.bajdcc.LALR1.interpret.module.web.ModuleNetWebServer
import com.bajdcc.util.ResourceLoader
import com.bajdcc.web.SpringBootstrap
import org.apache.log4j.Logger
import org.dom4j.Node
import org.dom4j.io.SAXReader
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.io.PrintWriter
import java.net.HttpURLConnection
import java.net.InetAddress
import java.net.URL
import java.net.UnknownHostException
import java.util.regex.Pattern

@Suppress("UNUSED_ANONYMOUS_PARAMETER")
/**
 * 【模块】网络
 *
 * @author bajdcc
 */
class ModuleNet : IInterpreterModule {
    private var runtimeCodePage: RuntimeCodePage? = null
    var server: ModuleNetServer? = null
    var client: ModuleNetClient? = null
    private var lastError = ""
    @get:Synchronized
    @set:Synchronized
    var webServer: ModuleNetWebServer? = null
    @get:Synchronized
    @set:Synchronized
    var webApi: ModuleNetWebApi? = null
    private var bootstrap: SpringBootstrap? = null

    override val moduleName: String
        get() = "sys.net"

    override val moduleCode: String
        get() = ResourceLoader.load(javaClass)

    override val codePage: RuntimeCodePage
        @Throws(Exception::class)
        get() {
            if (runtimeCodePage != null)
                return runtimeCodePage!!

            val base = ResourceLoader.load(javaClass)

            val grammar = Grammar(base)
            val page = grammar.codePage
            val info = page.info
            buildRemoteMethods(info)

            runtimeCodePage = page
            return page
        }

    private fun buildRemoteMethods(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_net_get",
                RuntimeDebugExec("HTTP GET", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        var text = ""
                        val txt = args[0].obj.toString()
                        logger.debug("Request url: $txt")
                        try {
                            val url = URL(txt)
                            val urlConnection = url.openConnection() // 打开连接
                            val br = BufferedReader(InputStreamReader(urlConnection.getInputStream(), "utf-8")) // 获取输入流
                            val sb = StringBuilder()
                            br.readLines().forEach { sb.append(it).append("\n") }
                            text = sb.toString()
                            br.close()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                        RuntimeObject(text)
                    }
                })
        info.addExternalFunc("g_net_get_json",
                RuntimeDebugExec("HTTP GET - json", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        var text = ""
                        val txt = args[0].obj.toString()
                        logger.debug("Request url(json): $txt")
                        try {
                            val url = URL(txt)
                            val urlConnection = url.openConnection() // 打开连接
                            urlConnection.setRequestProperty("accept", "*/*")
                            urlConnection.setRequestProperty("connection", "Keep-Alive")
                            urlConnection.setRequestProperty("user-agent", "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1)")
                            val br = BufferedReader(InputStreamReader(urlConnection.getInputStream(), "utf-8")) // 获取输入流
                            val sb = StringBuilder()
                            br.readLines().forEach { sb.append(it).append("\n") }
                            text = sb.toString()
                            br.close()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                        parseJson(text)
                    }
                })
        info.addExternalFunc("g_net_post_json",
                RuntimeDebugExec("HTTP POST - json", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        var text = ""
                        val txt = args[0].obj.toString()
                        val data = args[1].obj.toString()
                        logger.debug("Request url(json): $txt")
                        try {
                            val url = URL(txt)
                            val urlConnection = url.openConnection() as HttpURLConnection // 打开连接
                            urlConnection.doOutput = true
                            urlConnection.doInput = true
                            urlConnection.requestMethod = "POST"
                            urlConnection.setRequestProperty("accept", "*/*")
                            urlConnection.setRequestProperty("connection", "Keep-Alive")
                            urlConnection.setRequestProperty("user-agent", "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1)")
                            val out = PrintWriter(OutputStreamWriter(urlConnection.outputStream, "utf-8"))
                            out.print(data)
                            out.flush()
                            val br = BufferedReader(InputStreamReader(urlConnection.inputStream, "utf-8")) // 获取输入流
                            val sb = StringBuilder()
                            br.readLines().forEach { sb.append(it).append("\n") }
                            text = sb.toString()
                            br.close()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                        parseJson(text)
                    }
                })
        // -----------------------------------
        // server
        info.addExternalFunc("g_net_msg_create_server_internal",
                RuntimeDebugExec("MSG SERVER", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server != null || client != null)
                            RuntimeObject(false)
                        else {
                            val port = args[0].int
                            server = ModuleNetServer(port)
                            server!!.start()
                            RuntimeObject(true)
                        }
                    }
                })
        info.addExternalFunc("g_net_msg_shutdown_server",
                RuntimeDebugExec("MSG SERVER SHUTDOWN")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server == null)
                            RuntimeObject(false)
                        else {
                            server!!.exit()
                            RuntimeObject(true)
                        }
                    }
                })
        info.addExternalFunc("g_net_msg_get_server_status",
                RuntimeDebugExec("MSG SERVER GET STATUS")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server != null) {
                            val s = server!!.status
                            if (s == ModuleNetServer.Status.ERROR) {
                                lastError = server!!.error
                                server = null
                            }
                            status.service.processService.sleep(status.pid, MSG_QUERY_TIME)
                            RuntimeObject(s.ordinal.toLong())
                        } else RuntimeObject(ModuleNetServer.Status.NULL.ordinal.toLong())
                    }
                })
        // server
        // -----------------------------------
        // client
        info.addExternalFunc("g_net_msg_create_client_internal",
                RuntimeDebugExec("MSG CLIENT", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server != null || client != null)
                            RuntimeObject(false)
                        else {
                            val addr = args[0].obj.toString()
                            client = ModuleNetClient(addr)
                            client!!.start()
                            RuntimeObject(true)
                        }
                    }
                })
        info.addExternalFunc("g_net_msg_shutdown_client",
                RuntimeDebugExec("MSG CLIENT SHUTDOWN")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (client == null)
                            RuntimeObject(false)
                        else {
                            client!!.exit()
                            RuntimeObject(true)
                        }
                    }
                })
        info.addExternalFunc("g_net_msg_get_client_status",
                RuntimeDebugExec("MSG CLIENT GET STATUS")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (client != null) {
                            val s = client!!.status
                            if (s == ModuleNetClient.Status.ERROR) {
                                lastError = client!!.error
                                client = null
                            }
                            status.service.processService.sleep(status.pid, MSG_QUERY_TIME)
                            RuntimeObject(s.ordinal.toLong())
                        } else RuntimeObject(ModuleNetClient.Status.NULL.ordinal.toLong())
                    }
                })
        info.addExternalFunc("g_net_msg_client_send",
                RuntimeDebugExec("MSG CLIENT SEND", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (client != null) {
                            val s = client!!.status
                            if (s == ModuleNetClient.Status.RUNNING) {
                                status.service.processService.sleep(status.pid, MSG_SEND_TIME)
                                client!!.send(args[0].obj.toString())
                                RuntimeObject(true)
                            } else RuntimeObject(false)
                        } else RuntimeObject(false)
                    }
                })
        info.addExternalFunc("g_net_msg_client_send_with_origin",
                RuntimeDebugExec("MSG CLIENT SEND(ORIGIN)", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (client != null) {
                            val s = client!!.status
                            if (s == ModuleNetClient.Status.RUNNING) {
                                status.service.processService.sleep(status.pid, MSG_SEND_TIME)
                                client!!.send(args[0].obj.toString(), args[1].obj.toString())
                                RuntimeObject(true)
                            } else RuntimeObject(false)
                        } else RuntimeObject(false)
                    }
                })
        info.addExternalFunc("g_net_msg_server_send",
                RuntimeDebugExec("MSG SERVER SEND", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server != null) {
                            val s = server!!.status
                            if (s == ModuleNetServer.Status.RUNNING) {
                                status.service.processService.sleep(status.pid, MSG_SEND_TIME)
                                server!!.send(args[0].obj.toString())
                                RuntimeObject(true)
                            } else RuntimeObject(false)
                        } else RuntimeObject(false)
                    }
                })
        info.addExternalFunc("g_net_msg_server_send_error",
                RuntimeDebugExec("MSG SERVER SEND ERROR", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server != null) {
                            val s = server!!.status
                            if (s == ModuleNetServer.Status.RUNNING) {
                                status.service.processService.sleep(status.pid, MSG_SEND_TIME)
                                server!!.send_error(args[0].obj.toString())
                                RuntimeObject(true)
                            } else RuntimeObject(false)
                        } else RuntimeObject(false)
                    }
                })
        info.addExternalFunc("g_net_msg_server_send_with_origin",
                RuntimeDebugExec("MSG SERVER SEND(ORIGIN)", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        if (server != null) {
                            val s = server!!.status
                            if (s == ModuleNetServer.Status.RUNNING) {
                                status.service.processService.sleep(status.pid, MSG_SEND_TIME)
                                server!!.send(args[0].obj.toString(), args[1].obj.toString())
                                RuntimeObject(true)
                            } else RuntimeObject(false)
                        } else RuntimeObject(false)
                    }
                })
        // client
        // -----------------------------------
        info.addExternalFunc("g_net_msg_get_error",
                RuntimeDebugExec("MSG SERVER GET ERROR")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(lastError) })
        info.addExternalFunc("g_net_msg_get_server_msg",
                RuntimeDebugExec("MSG SERVER GET MESSAGE")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(if (server == null) null else server!!.message) })
        info.addExternalFunc("g_net_msg_get_client_msg",
                RuntimeDebugExec("MSG CLIENT GET MESSAGE")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(if (client == null) null else client!!.message) })
        info.addExternalFunc("g_net_msg_get_client_addr",
                RuntimeDebugExec("MSG CLIENT GET ADDRESS")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(if (client == null) null else client!!.addr) })
        info.addExternalFunc("g_net_parse_json",
                RuntimeDebugExec("Parse json", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus -> parseJsonSafe(args[0].obj.toString()) })
        info.addExternalFunc("g_net_get_rss",
                RuntimeDebugExec("RSS GET(BAIDU)", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val txt = args[0].obj.toString()
                        logger.debug("Request url(rss): $txt")
                        val array = RuntimeArray()
                        val pattern = Pattern.compile("<br>(.*)<br />")
                        try {
                            val reader = SAXReader()
                            val document = reader.read(txt)
                            val title = document.selectSingleNode("//title")
                            array.add(RuntimeObject(title.text))
                            val list = document.selectNodes("//item").map { it as Node }
                            for (item in list) {
                                val itemTitle = item.valueOf("title")
                                array.add(RuntimeObject(itemTitle))
                                val itemDescription = item.valueOf("description")
                                val matcher = pattern.matcher(itemDescription)
                                if (matcher.find()) {
                                    array.add(RuntimeObject(matcher.group(1)))
                                } else {
                                    array.add(RuntimeObject(itemDescription.replace("<br />", "")))
                                }
                            }
                            RuntimeObject(array)
                        } catch (e: Exception) {
                            e.printStackTrace()
                            RuntimeObject(array)
                        }
                    }
                })
        info.addExternalFunc("g_net_start_web",
                RuntimeDebugExec("WEB SERVER", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        try {
                            val text = args[0].obj.toString()
                            val port = Integer.parseInt(text)
                            Thread(ModuleNetWebServer(port)).start()
                            RuntimeObject(true)
                        } catch (e: Exception) {
                            e.printStackTrace()
                            RuntimeObject(false)
                        }
                    }
                })
        info.addExternalFunc("g_net_stop_web",
                RuntimeDebugExec("STOP WEB SERVER")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val server = webServer
                        if (server != null) {
                            server.isRunning = false
                            webServer = null
                            RuntimeObject(true)
                        } else {
                            RuntimeObject(false)
                        }
                    }
                })
        info.addExternalFunc("g_net_query_web",
                RuntimeDebugExec("QUERY WEB SERVER")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(webServer != null) })
        info.addExternalFunc("g_net_has_request",
                RuntimeDebugExec("WEB SERVER - REQUEST")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val server = webServer
                        if (server != null) {
                            RuntimeObject(server.hasRequest())
                        } else {
                            RuntimeObject(false)
                        }
                    }
                })
        info.addExternalFunc("g_net_get_request",
                RuntimeDebugExec("WEB SERVER - PEEK REQUEST")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val server = webServer
                        if (server != null) {
                            RuntimeObject(server.peekRequest())
                        } else {
                            RuntimeObject(false)
                        }
                    }
                })
        info.addExternalFunc("g_web_get_ip",
                RuntimeDebugExec("获取IP地址")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        try {
                            val address = InetAddress.getLocalHost()
                            RuntimeObject(address.hostAddress)
                        } catch (e: UnknownHostException) {
                            e.printStackTrace()
                            RuntimeObject("Unknown IP Address")
                        }
                    }
                })
        info.addExternalFunc("g_web_get_hostname",
                RuntimeDebugExec("获取主机名")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        try {
                            val address = InetAddress.getLocalHost()
                            RuntimeObject(address.hostName)
                        } catch (e: UnknownHostException) {
                            e.printStackTrace()
                            RuntimeObject("Unknown Hostname")
                        }
                    }
                })
        info.addExternalFunc("g_web_start_spring",
                RuntimeDebugExec("启动Spring Boot")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        webApi = ModuleNetWebApi()
                        bootstrap = SpringBootstrap()
                        bootstrap!!.start()
                        null
                    }
                })
        info.addExternalFunc("g_web_stop_spring",
                RuntimeDebugExec("关闭Spring Boot")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        webApi = null
                        if (bootstrap != null) {
                            bootstrap!!.terminate()
                            bootstrap!!.join()
                        }
                        null
                    }
                })
    }

    @Synchronized
    fun resetBootstrap() {
        this.bootstrap = null
    }

    companion object {

        private val logger = Logger.getLogger("net")
        val instance = ModuleNet()

        private const val MSG_QUERY_TIME = 15
        private const val MSG_SEND_TIME = 5

        private fun parseJsonSafe(text: String): RuntimeObject? {
            try {
                val o = JSON.parseObject(text)
                return parseInternal(o)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return null
        }

        private fun parseJson(text: String): RuntimeObject {
            try {
                val o = JSON.parseObject(text)
                return parseInternal(o)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return RuntimeObject("json - parse error")
        }

        private fun parseInternal(o: Any): RuntimeObject {
            if (o is JSONObject) {
                val map = RuntimeMap()
                o.forEach { key, value -> map.put(key, parseInternal(value)) }
                return RuntimeObject(map)
            } else if (o is JSONArray) {
                val arr = RuntimeArray()
                o.forEach { key -> arr.add(parseInternal(key)) }
                return RuntimeObject(arr)
            } else {
                if (o is Int) {
                    return RuntimeObject(o.toLong())
                } else if (o is Float) {
                    return RuntimeObject(o.toDouble())
                }
                return RuntimeObject(o)
            }
        }
    }
}
